Actual Costs - API and UI support

mrsheepuk opened the pull request:
## Summary

This introduces the persistence layer for asset costs into Kore, adds API endpoints to report on those actual costs, and provides a simple UI to show the overall summary and team summary.

**Which issue(s) this PR resolves**:
Resolves #1156

## Checklist

- [ ] ~~I've created a PR to update the [documentation](https://github.com/appvia/kore-docs):~~ - documentation updates will come with the next PR to enable cost providers and make the end-to-end costs flow work.

## Changes

* Add new costs models to the persistence layer with appropriate functions to access it.
* Add new API types for cost summaries overall, by team, and by asset.
* Add UI to show overall and team summaries.

### Additional changes

- Ensure that all clusters are additionally tagged with a kore-instance tag, using the identifier assigned to the kore-admin team as the unique identifier for an instance of kore. This is to allow multiple instances of kore to be used against a single cloud account without them confusing each other.
- Add makefile target to quickly exec into the mysql container when running in Kind.
- Add support for setting the db layer to verbose logging in the helm chart.

## Screenshots

### Summary:

<img width="1282" alt="kore-costs-summary" src="https://user-images.githubusercontent.com/7505593/89290839-5f6e9880-d651-11ea-86eb-b7cb96779da9.png">


### Team drill-down:

<img width="1282" alt="kore-costs-team" src="https://user-images.githubusercontent.com/7505593/89290853-65fd1000-d651-11ea-9407-67fe6f8372a1.png">

### Team drill-down with asset detail:

<img width="1282" alt="kore-costs-team-assets" src="https://user-images.githubusercontent.com/7505593/89290888-73b29580-d651-11ea-937a-73eb06c838b7.png">

### All-in-one if your screen is wide enough:

![image](https://user-images.githubusercontent.com/7505593/89291052-bffdd580-d651-11ea-859c-eb6b885d0673.png)

## Testing

This has been extensively tested by generating some real resources then manually importing cost data into Kore using the new API. Once some cost providers are written, testing the end-to-end flows will be possible.

## Follow-up stories

Stories have been raised for the GCP, AWS and Azure cost providers:
- appvia/kore-planning#147
- appvia/kore-planning#148
- appvia/kore-planning#149

Other Kore stories:
- Team member cost view #1203

davet1985 commented on ui/lib/components/costs/OverallCostsViewer.js:
```diff
@@ -0,0 +1,130 @@
+import { getCloudInfo } from '../../utils/cloud'
+
+export default class OverallCostsViewer extends React.Component {
+  static propTypes = {
+  }
```
Remove this and the PropTypes import?

davet1985 commented on ui/lib/components/costs/OverallCostsViewer.js:
```diff
@@ -0,0 +1,130 @@
+    summary: null,
+    loading: false,
+    provider: '',
+    from:  moment().utc(false).startOf('month'),
```
nitpick - double space

davet1985 commented on ui/lib/components/costs/OverallCostsViewer.js:
```diff
@@ -0,0 +1,130 @@
+
+  state = {
+    summary: null,
+    loading: false,
```
Should `loading` be set to `true` when the component is first created?

davet1985 commented on ui/lib/components/costs/OverallCostsViewer.js:
```diff
@@ -0,0 +1,130 @@
+
+  prevMonth = () => {
+    let { from, to } = this.state
+    from = moment(from).utc(false).add(-1, 'month').startOf('month')
```
These moment chains are hard to read, could they be pulled to somewhere central, either in this file or another and given sensible names to help understanding?

mrsheepuk replied:
refactored - take a look d9921dd0

davet1985 replied:
Looks much better 👍

davet1985 commented on ui/lib/components/costs/OverallCostsViewer.js:
```diff
@@ -0,0 +1,130 @@
+            <Card title="Summary">
+              {error ? <Alert type="error" message={error.message} /> : null}
+              <OverallCostSummary 
+                // style={{ marginTop: '20px' }} 
```
Remove if not required

bandesz commented on pkg/costs/assets.go:
```diff
@@ -0,0 +1,279 @@
+			return err
+		}
+	}
+	for ind, modelCost := range modelCosts {
```
Here we really need a transaction, otherwise we can save partial data. As there is no unique index on team_asset_costs, multiple calls will save the same partial data multiple times.

mrsheepuk replied:
agreed, fixed in b7a6ba2 - and unique identifier added too (with a default population for now if the cost provider doesn't provide one)

bandesz commented on pkg/costs/assets.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+type assetsImpl struct {
+	getKoreIdentifier func() string
```
it seems this is only used from ListAssets, so maybe it would be nicer if we pass the identifier value as a parameter to ListAssets?

mrsheepuk replied:
I disagree - it's an implementation detail of how the assets are constructed, callers shouldn't need to know that.

bandesz commented on pkg/kore/clusters.go:
```diff
@@ -306,10 +309,16 @@ func (c *clustersImpl) ensureIdentifiers(ctx context.Context, cluster *clustersv
+	// work out the right cloud for the cluster provider so we can tag the asset correctly
+	cloud, mapErr := c.Costs().Metadata().MapProviderToCloud(cluster.Spec.Kind)
+	if mapErr != nil {
+		log.WithField("provider", cluster.Spec.Kind).WithError(mapErr).Warn("received error mapping k8s provider to cloud provider for asset identifier, will continue")
```
I don't think we should swallow this error. If we introduce a new cloud provider, this will help us to catch a missing mapping right away.

mrsheepuk replied:
fixed in a3336ce

bandesz commented on pkg/kore/setup.go:
```diff
@@ -384,22 +393,29 @@ func (h hubImpl) ensureTeamsAndClustersIdentified(ctx context.Context) error {
-					cluster.Labels = map[string]string{}
+				cloud, mapErr := h.Costs().Metadata().MapProviderToCloud(cluster.Spec.Kind)
+				if mapErr != nil {
+					log.WithField("provider", cluster.Spec.Kind).WithError(mapErr).Warn("received error mapping k8s provider to cloud provider for asset identifier, will continue with blank cloud provider")
```
we shouldn't swallow this error either.

mrsheepuk replied:
there's at least one known case where this will be nil - the kore cluster doesn't have a cloud. I think in the setup.go, we shouldn't explode if this isn't available, the warning should be enough.

bandesz replied:
We could still map it to something internal, e.g. "kore", "unmanaged". Nobody will ever see that warning

mrsheepuk replied:
fixed in a3336ce

bandesz commented on pkg/persistence/model/migration.go:
```diff
@@ -91,9 +91,13 @@ func Migrations(db *gorm.DB) error {
-		AddForeignKey("team_identifier", "team_identity(team_identifier)", "RESTRICT", "RESTRICT").
-		AddUniqueIndex("idx_asset", "team_identifier", "asset_identifier").
-		AddIndex("idx_assettype", "team_identifier", "asset_type")
+		AddIndex("idx_team", "team_identifier").
```
There is no primary key on this table, we should add back asset_identifier probably? Also the model has that field marked as primary key too.

mrsheepuk replied:
I suspect I've got my gorm mixed up - I was under the impression that defining it on the model would automatically cause the primary key to be created on the table as the model defines, but perhaps I was wrong. Will double check the created DB objects and rectify as needed.

mrsheepuk replied:
I've verified this - the `gorm:"primary_key"` annotation sets the primary key on the database objects, don't need to put them on the migration. I've added the foreign keys back in though, as they ARE needed.

mrsheepuk replied:
foreign keys added back in - b7a6ba2

bandesz replied:
I've checked the `SHOW CREATE TABLE` directly and there was no primary key.

mrsheepuk replied:
As discussed, verified this directly from a clean database and it does create the primary keys - perhaps there's an issue with the migration?

I'd suggest we should merge this and check what happens when this is deployed onto other environments, as if this isn't creating the primary keys correctly it may well affect other tables as well, so we'd need to fix generally rather than just for this case.

mrsheepuk replied:
Resolved in 25e87f1

bandesz commented on pkg/persistence/model/team_asset.go:
```diff
@@ -39,18 +39,24 @@ const (
 // cloud service, etc) owned by that team - can be used for (e.g.) cost tracking etc.
 type TeamAsset struct {
+	// AssetIdentifier is the identity of the asset in question
+	AssetIdentifier string `sql:"type:char(20)" gorm:"primary_key"`
```
I'm not sure this matters, but there is a discrepancy here, that this field is marked as a primary key, but it's not one in the db.

mrsheepuk replied:
as above - this does set the primary key

bandesz commented on pkg/persistence/model/team_asset_cost.go:
```diff
@@ -0,0 +1,55 @@
+	// Account indicates which account / project / subscription this cost relates to
+	Account string
+	// BillingYear is the (4-digit) year in which this cost was billed (e.g. 2020)
+	BillingYear uint16
```
We do we need these as separate fields? What is the value of these if I query a time range containing multiple months?

mrsheepuk replied:
all three supported clouds have a conceptual billing / invoice month - they are not necessarily the same month as when the usage occurs. by replicating these values through into the data model, it would support any future use cases for grouping by invoice (instead of by arbitrary time range over when the usage occurred). 

that said, I don't have any idea why I've made them two separate fields, now I look at it. will combine.

mrsheepuk replied:
done - b7a6ba2

bandesz commented on pkg/persistence/team_assets.go:
```diff
@@ -0,0 +1,222 @@
+
+// RecordTeamIdentiy records the existence of a team
+func (t teamAssetsImpl) RecordTeamIdentity(ctx context.Context, teamIdentifier string, teamName string) error {
+	timed := prometheus.NewTimer(setLatency)
```
Unrelated to the PR, but we should really have a layer over the DB object which records these automatically

bandesz commented on pkg/persistence/team_assets_filter.go:
```diff
@@ -0,0 +1,119 @@
+// NewTeamAssetListOptions returns a cost list options
+func NewTeamAssetListOptions() *TeamAssetListOptions {
+	return &TeamAssetListOptions{
+		WithDeleted: false,
```
This is the default empty value, so no need to define it.

mrsheepuk replied:
Yeah - I realised this, but I wanted to be explicit about the intention - so it's clear to anyone who looks at this code that the intention is to hide deleted by default (rather than by accident). 

Helps poor souls wondering in future why there's no deleted records in their query (poor souls like I was when the deleted rows weren't there because gorm silently hides rows with a column called deleted_at if that column is populated on a row...)

mrsheepuk commented on charts/kore/templates/kore.yml:
```diff
@@ -139,6 +139,8 @@ spec:
             value: "{{ .Values.api.enable_metrics }}"
           - name: METRICS_PORT
             value: "{{ .Values.api.metrics_port }}"
+          - name: ENABLE_USER_DB_LOGGING
```
I didn't introduce that env variable, just added support for setting it from the helm chart... agree it would be good to rename though. Will take a look.

mrsheepuk replied:
I've just checked and all the env variables / parameters call this the user db (I suspect because when it was first introduced, that's all it stored). think renaming those should be a separate PR as very not related to this PR.

mrsheepuk commented on pkg/apis/costs/v1beta1/cost_types.go:
```diff
@@ -17,53 +17,136 @@
-// CostList represents a collection of costs
+// CostAsset represents a resource known to Kore which a cost provider should provide costs data for
+// +k8s:openapi-gen=false
+type CostAsset struct {
```
done in b7a6ba2

mrsheepuk commented:
@bandesz Addressed your comments, also added an extra commit to promote the metadata bit out of the costs package into its own top-level package as previously suggested (I was messing with that code anyway)

a3336ce - move metadata
b7a6ba2 - address review comments (backend)
ac42c40 - small UI improvements (cc @davet1985 if you want to take another look)

If you could give it a final once-over and approve, I'd appreciate it.

mrsheepuk commented:
@bandesz Added an explicit database migration to address the primary key issue in 25e87f1 - just tested with both a clean 'up' and a 'master-up-branch-run' and in both cases, the primary key now ends up on the table.
